flow/aggregates.py

```diff
@@ -0,0 +1,475 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Iterable
+from hashlib import md5
+from itertools import groupby
+from itertools import zip_longest
+
+
+class aggregator:
+    r"""Decorator for operation function that is to be aggregated.
+    By default, if the ``aggregator_function`` is not passed,
+    an aggregate of all jobs will be created.
+
+    .. code-block:: python
+
+        @aggregator()
+        @FlowProject.operation
+        def foo(*jobs):
+            print(len(jobs))
+
+    :param aggregator_function:
+        A callable that performs aggregation of jobs. It takes in a list of
+        jobs and can return or yield subsets of jobs as an iterable. The
+        default behavior is creating a single aggregate of all jobs.
+    :type aggregator_function:
+        callable
+    :param sort_by:
+        Before aggregating, sort the jobs by a given statepoint parameter.
+        The default behavior is no sorting.
+    :type sort_by:
+        str or NoneType
+    :param reverse_order:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse_order:
+        bool
+    :param select:
+        Condition for filtering individual jobs. This is passed as the
+        callable argument to `filter`. The default behavior is no filtering.
+    :type select:
+        callable or NoneType
+    :param tag:
+        Additional information about the aggregator object which is a factor of
+        difference between two aggregator objects. signac-flow uses byte code
+        to compare callables like `aggregator_function` and `select`. Since a
+        byte code cannot truly differentiate between callables hence the tag,
+        if provided, becomes a factor of differentiation.
```

Comment:
Does setting tags affect correctness, or is it just a performance optimization? Now I'm actually a little confused what the goal of the tags are. Are there cases where a user would have to provide tags, otherwise we might think that two aggregates are equivalent and therefore reuse the same set of job aggregates for them even though they should actually be using different ones? That seems _highly_ problematic to me. If the worst case scenario is just that we regenerate too many aggregates, then providing tags as a way for users to speed that up is fine, but we should never be in a situation where we silently use the wrong aggregates.

It should be possible to avoid any false positives with equality, it's just not possible to avoid getting false negatives (which is where the optimization comes in). Can you give an example where the current test fails to distinguish functions without the use of tags? Does that issue get fixed if you also check `co_consts` in addition to `co_code`?

Reply:
I did not try `co_consts`. I will try to make that work and let you know

Reply:
Well first, can you give an example where the tag is _necessary_, i.e. we would think two aggregators are identical when they really aren't? I have no idea if `co_consts` would fix the problem because I don't know what it is, I just suggested that because there is more information available than `co_code` if that is insufficient to distinguish.

For a feature like this we _cannot_ rely on the user to provide tags to prevent incorrectly identifying two distinct aggregates as the same. If we really can't distinguish them for some reason, I think we would have to eat the performance cost and just copy all aggregates and not reuse any of them. Optimizations cannot result in incorrect behavior, and we can't push that responsibility onto the user.

Reply:
When we use `co_code`, we are not able to differentiate between `groupsof(2)` and `groupsof(3)`.
I believe this is because the function doesn't itself store any information about the number provided.
It just references to the value stored for num.
This is where `tag` comes handy. `co_consts` will not work here because it will just provide the constants declared in the function (which we don't do).

Reply:
Yes that's what I would expect. In this particular case you can get around the problem by instead looking at `x.__closure__`. However, I'm not sure if this would work in *all* cases. I think if you traversed all members of `x.__code__` and `x.__closure__` and checked them each for equality with the same members of another function you could guarantee that you don't get false positive, but I'm not 100% sure about that. For this particular use case, I think guaranteeing 0 false positives is far more important than preventing false negatives; the latter is just a performance hit, whereas the former leads to incorrect behavior. Reiterating what I said in my previous comment, it's much better to store some extra aggregates than to use the wrong aggregates for operations.

@b-butler what are your thoughts here?

Reply:
I was not aware of `__closure__`, but that would likely be sufficient then. I don't care about false negatives, but yes we need to be sure of no false positives. We should also be fairly sure that `groupsof` and `groupby` are not giving false negatives.